plugins/obs-ffmpeg/obs-ffmpeg-source.c

```diff
@@ -87,20 +87,73 @@ static bool is_local_file_modified(obs_properties_t *props,
 	obs_property_t *input_format =
 		obs_properties_get(props, "input_format");
 	obs_property_t *local_file = obs_properties_get(props, "local_file");
-	obs_property_t *looping = obs_properties_get(props, "looping");
+	obs_property_t *end_playback =
+		obs_properties_get(props, "end_playback");
 	obs_property_t *buffering = obs_properties_get(props, "buffering_mb");
 	obs_property_t *seekable = obs_properties_get(props, "seekable");
 	obs_property_t *speed = obs_properties_get(props, "speed_percent");
 	obs_property_t *reconnect_delay_sec =
 		obs_properties_get(props, "reconnect_delay_sec");
+	obs_property_t *clear_on_media_end =
+		obs_properties_get(props, "clear_on_media_end");
 	obs_property_set_visible(input, !enabled);
 	obs_property_set_visible(input_format, !enabled);
 	obs_property_set_visible(buffering, !enabled);
 	obs_property_set_visible(local_file, enabled);
-	obs_property_set_visible(looping, enabled);
+	obs_property_set_visible(end_playback, enabled);
 	obs_property_set_visible(speed, enabled);
 	obs_property_set_visible(seekable, !enabled);
 	obs_property_set_visible(reconnect_delay_sec, !enabled);
+	obs_property_set_visible(clear_on_media_end, !enabled);
+
+	return true;
+}
+
+static bool is_end_playback_modified(obs_properties_t *props,
+				     obs_property_t *prop, obs_data_t *settings)
+{
+	UNUSED_PARAMETER(prop);
+	UNUSED_PARAMETER(props);
+	int end_playback = obs_data_get_int(settings, "end_playback");
+	bool looping = false;
+	bool clear_on_media_end = true;
+
+	if (end_playback == 1) { // End action : Hide
+		clear_on_media_end = false;
+	} else if (end_playback == 2) { // End action : Loop
+		clear_on_media_end = false;
+		looping = true;
+	}
+	obs_data_set_bool(settings, "looping", looping);
+	obs_data_set_bool(settings, "clear_on_media_end", clear_on_media_end);
+
+	return true;
+}
+
+static void update_end_playback_property_list(obs_data_t *settings)
+{
+	bool clear_on_media_end =
+		obs_data_get_bool(settings, "clear_on_media_end");
+	bool looping = obs_data_get_bool(settings, "looping");
+	int end_playback = 0;
+
+	if (looping == true) {
+		end_playback = 2;
+	} else if (clear_on_media_end == false) {
+		end_playback = 1;
+	}
+
+	obs_data_set_int(settings, "end_playback", end_playback);
```

Comment:
You probably also want to use `obs_data_erase` to remove the old setting when upgrading it to the new variant. Otherwise the old setting will persist in a user's configuration and you will always go and update it.

Probably worthwhile to check if the old setting exists first and only go through the effort to upgrade when it's necessary in the first place.

Reply:
A "safer" method would be to only upgrade the setting if the old setting exists and the new setting does not. Erasing old setting values completely breaks configs for older versions.

Reply:
Isn't that expected? If I were to create the source anew after the change, I could not go back to an older version of OBS and expect the setting to be "downgraded" either.

I'm aware of programs commonly upgrading settings but not that these upgraded settings are still expected to work with older versions (new settings would just be ignored).

Reply:
I'm not saying that the upgraded settings should work with old versions. I'm saying that if you open a newer OBS one time that deletes settings entirely, and then you open an older OBS because you only wanted to run the newer OBS once to check it out, your settings are gone. I'm aware that we don't want to encourage or support downgrading, but purposefully breaking stuff seems a bit risky for users who just want to test stuff (and don't know that we encourage using portable mode or backing up your settings beforehand).

Reply:
> Fine with me, there's no issue with keeping it. The property should be removed nevertheless though.

Agree with that.